src/libData/AccountData/AccountStoreSC.tpp

```diff
@@ -442,45 +445,66 @@ bool AccountStoreSC<MAP>::UpdateAccounts(const uint64_t& blockNum,
       }
 
       // Check if the sender has enough balance to pay gasDeposit
-      if (fromAccount->GetBalance() < gasDeposit) {
+      if (fromAccount->GetBalance() < gasDeposit + transaction.GetAmount()) {
         LOG_GENERAL(WARNING,
                     "The account doesn't have enough gas to create a contract");
         error_code = TxnStatus::INSUFFICIENT_BALANCE;
         return false;
       }
-      // deduct scilla checker invoke gas
-      if (gasRemained < SCILLA_CHECKER_INVOKE_GAS) {
-        LOG_GENERAL(WARNING, "Not enough gas to invoke the scilla checker");
-        error_code = TxnStatus::INSUFFICIENT_GAS;
-        return false;
-      } else {
-        gasRemained -= SCILLA_CHECKER_INVOKE_GAS;
+
+      if (isScilla) {
+        // deduct scilla checker invoke gas
+        if (gasRemained < SCILLA_CHECKER_INVOKE_GAS) {
+          LOG_GENERAL(WARNING, "Not enough gas to invoke the scilla checker");
+          error_code = TxnStatus::INSUFFICIENT_GAS;
+          return false;
+        } else {
+          gasRemained -= SCILLA_CHECKER_INVOKE_GAS;
+        }
       }
 
       // generate address for new contract account
       Address contractAddress =
           Account::GetAddressForContract(fromAddr, fromAccount->GetNonce());
       // instantiate the object for contract account
       // ** Remeber to call RemoveAccount if deployment failed halfway
-      if (!this->AddAccount(contractAddress, {0, 0})) {
-        LOG_GENERAL(WARNING, "AddAccount failed for contract address "
-                                 << contractAddress.hex());
-        error_code = TxnStatus::FAIL_CONTRACT_ACCOUNT_CREATION;
-        return false;
-      }
-      Account* contractAccount = this->GetAccount(contractAddress);
-      if (contractAccount == nullptr) {
-        LOG_GENERAL(WARNING, "contractAccount is null ptr");
-        error_code = TxnStatus::FAIL_CONTRACT_ACCOUNT_CREATION;
-        return false;
+      Account* contractAccount;
```

Comment:
This whole Atomics Area of the Accountstore is another obscure area of code that needs some clarity, should be addressed in a future refactoring.